File: 백준/Gold/9935. 문자열 폭발/문자열 폭발.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    private static String original;
    private static String bomb;

    public static void main(String[] args) throws IOException {
        init();
        findAnswer();
    }

    private static void init() throws IOException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        original = input.readLine();
        bomb = input.readLine();
    }

    private static void findAnswer() {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < original.length(); i++) {
            stack.push(original.charAt(i));

            if (stack.size() >= bomb.length()) {
                boolean flag = true;
                for (int j = 0; j < bomb.length(); j++) {
                    if (bomb.charAt(j) != stack.get(stack.size() - bomb.length() + j)) {
                        flag = false;
                        break;
                    }
                }
                if (flag) {
                    for (int j = 0; j < bomb.length(); j++) {
                        stack.pop();
                    }
                }
            }
        }
        StringBuilder answer = new StringBuilder();
        if (stack.isEmpty()) {
            answer.append("FRULA");
        } else {
            for (char c : stack) {
                answer.append(c);
            }
        }

        System.out.println(answer);
    }
}

File: 백준/Gold/9935. 문자열 폭발/문자열 폭발.cc
#include <bits/stdc++.h>
using namespace std;
string s;
string bomb;
stack<char> st;
string ret;
void go() {
    string temp;
    int cnt = 0;
    for (int i = (int)bomb.size() - 1; i >= 0; i--) {
        if (bomb[i] == st.top()) {
            cnt++;
            temp+=st.top();
            st.pop();
        }
    }

    if (cnt == (int) bomb.size()) {
        return;
    } else {
        // 다시 거꾸로 넣어주자
        for (int i = (int)temp.size() - 1; i >= 0; i--) {
            st.push(temp[i]);
        }
    }
}
int main(){
    cin >> s >> bomb;
    if ((int)s.size() < (int)bomb.size()) {
        cout << s << '\n';
        return 0 ;
    }
    
    for (int i = 0; i < (int) s.size(); i++) {
        char c = s[i];
        st.push(c);

        if (st.size() && st.top() == bomb[bomb.size() - 1] && (int)st.size() >= (int)bomb.size()) {
            go();
        }
    }

    if(st.empty()) {
        cout << "FRULA" << '\n';
        return 0;
    }

    while (st.size()) {
        ret += st.top();
        st.pop();
    }
    
    reverse(ret.begin(), ret.end());

    cout << ret << '\n';
    return 0;
}